File: src/test/test.selectiontotale.c
#include "selectiontotale.h"
#include "test_utils.h"
#include <stdlib.h>

int main(int argc, char** argv)
{
	int* dimensions = readArray_int("dimensions");
	int n = dimensions[0];
	int p = dimensions[1];
	int m = dimensions[2];
	int k = dimensions[3];
	int L = dimensions[4];
	free(dimensions);

	////////////
	// INPUTS //
	Real* phiInit = readArray_real("phiInit");
	Real* rhoInit = readArray_real("rhoInit");
	Real* piInit = readArray_real("piInit");
	Real* gamInit = readArray_real("gamInit");
	int mini = read_int("mini");
	int maxi = read_int("maxi");
	Real gamma = read_real("gamma");
	Real* glambda = readArray_real("glambda");
	Real* X = readArray_real("X");
	Real* Y = readArray_real("Y");
	Real seuil = read_real("seuil");
	Real tau = read_real("tau");
	////////////

	/////////////
	// OUTPUTS //
	int* A1 = (int*)malloc(p*(m+1)*L*sizeof(int));
	int* A2 = (int*)malloc(p*(m+1)*L*sizeof(int));
	Real* Rho = (Real*)malloc(m*m*k*L*sizeof(Real));
	Real* Pi = (Real*)malloc(k*L*sizeof(Real));
	/////////////

	/////////////////////////////////////////
	// Call to constructionModelesLassoMLE //
	selectiontotale_core(
		phiInit,rhoInit,piInit,gamInit,mini,maxi,gamma,glambda,X,Y,seuil,tau,
		A1,A2,Rho,Pi,
		n,p,m,k,L);
	/////////////////////////////////////////

	free(phiInit);
	free(rhoInit);
	free(piInit);
	free(gamInit);
	free(glambda);
	free(X);
	free(Y);

	// Compare to reference outputs
	int* ref_A1 = readArray_int("A1");
	compareArray_int("A1", A1, ref_A1, p*(m+1)*L);
	free(A1);
	free(ref_A1);

	int* ref_A2 = readArray_int("A2");
	compareArray_int("A2", A2, ref_A2, p*(m+1)*L);
	free(A2);
	free(ref_A2);

	Real* ref_Rho = readArray_real("Rho");
	compareArray_real("Rho", Rho, ref_Rho, m*m*k*L);
	free(Rho);
	free(ref_Rho);

	Real* ref_Pi = readArray_real("Pi");
	compareArray_real("Pi", Pi, ref_Pi, k*L);
	free(Pi);
	free(ref_Pi);

	return 0;
}

File: src/test/OLD_TEST_MATLAB/testSelectiontotale.m
function[] = testSelectiontotale()

	testFolder = 'data/';
	delimiter = '\n';

	%get dimensions
	dimensions = dlmread(strcat(testFolder,'dimensions'), delimiter);
	n = dimensions(1);
	p = dimensions(2);
	m = dimensions(3);
	k = dimensions(4);
	L = dimensions(5);

	%get all input arrays
	phiInit = reshape(dlmread(strcat(testFolder,'phiInit'), delimiter), p, m, k);
	rhoInit = reshape(dlmread(strcat(testFolder,'rhoInit'), delimiter), m, m, k);
	piInit = transpose(dlmread(strcat(testFolder,'piInit'), delimiter));
	gamInit = reshape(dlmread(strcat(testFolder,'gamInit'), delimiter), n, k);
	mini = int64(dlmread(strcat(testFolder,'mini'), delimiter));
	maxi = int64(dlmread(strcat(testFolder,'maxi'), delimiter));
	gamma = dlmread(strcat(testFolder,'gamma'), delimiter);
	glambda = dlmread(strcat(testFolder,'glambda'), delimiter);
	X = reshape(dlmread(strcat(testFolder,'X'), delimiter), n, p);
	Y = reshape(dlmread(strcat(testFolder,'Y'), delimiter), n, m);
	seuil = dlmread(strcat(testFolder,'seuil'), delimiter);
	tau = dlmread(strcat(testFolder,'tau'), delimiter);

	%run constructionModelesLassoMLE.m
	[A1,A2,Rho,Pi] = selectiontotale(phiInit,rhoInit,piInit,gamInit,mini,maxi,gamma,glambda,X,Y,seuil,tau);

	%get all stored outputs
	ref_A1 = int64(reshape(dlmread(strcat(testFolder,'A1'), delimiter), p, m+1, L));
	ref_A2 = int64(reshape(dlmread(strcat(testFolder,'A2'), delimiter), p, m+1, L));
	ref_Rho = reshape(dlmread(strcat(testFolder,'Rho'), delimiter), m, m, k, L);
	ref_Pi = reshape(dlmread(strcat(testFolder,'Pi'), delimiter), k, L);

	%check that output correspond to stored output
	tol = 1e-5;
	checkOutput('A1',A1,ref_A1,tol);
	checkOutput('A2',A2,ref_A2,tol);
	checkOutput('Rho',Rho,ref_Rho,tol);
	checkOutput('Pi',Pi,ref_Pi,tol);

end
